src/ctap/credential_id.rs

```diff
@@ -0,0 +1,254 @@
+    PrivateKey = 0,
+    RpIdHash = 1,
+    CredProtectPolicy = 2,
+}
+
+impl From<CborKeyHandleField> for sk_cbor::Value {
+    fn from(field: CborKeyHandleField) -> sk_cbor::Value {
+        (field as u64).into()
+    }
+}
+
+fn decrypt_cbor_key_handle(
+    env: &mut impl Env,
+    bytes: &[u8],
+) -> Result<Option<CredentialId>, Ctap2StatusCode> {
+    let aes_enc_key = crypto::aes256::EncryptionKey::new(&env.key_store().key_handle_encryption()?);
+    let mut plaintext = aes256_cbc_decrypt(&aes_enc_key, bytes, true)?;
+    remove_padding(&mut plaintext)?;
+
+    let decoded_cbor = cbor_read(plaintext.as_slice())?;
+    destructure_cbor_map! {
+      let {
+          CborKeyHandleField::PrivateKey => private_key,
+          CborKeyHandleField::RpIdHash=> rp_id_hash,
+          CborKeyHandleField::CredProtectPolicy => cred_protect_policy,
+      } = extract_map(decoded_cbor)?;
+    }
+    Ok(match (private_key, rp_id_hash, cred_protect_policy) {
+        (Some(private_key), Some(rp_id_hash), cred_protect_policy) => {
+            let private_key = PrivateKey::try_from(private_key)?;
+            let rp_id_hash = extract_byte_string(rp_id_hash)?;
+            let cred_protect_policy = if let Some(policy) = cred_protect_policy {
+                Some(CredentialProtectionPolicy::try_from(policy)?)
+            } else {
+                None
+            };
+            Some(CredentialId {
+                private_key,
+                rp_id_hash,
+                cred_protect_policy,
+            })
+        }
+        _ => None,
+    })
+}
+
+fn is_legacy_key_handle_version(version: u8) -> bool {
+    version == ECDSA_CREDENTIAL_ID_VERSION || version == ED25519_CREDENTIAL_ID_VERSION
+}
+
+/// Pad data to MAX_PADDING_LENGTH+1 (192) bytes using PKCS padding scheme.
+/// Let N = 192 - data.len(), the PKCS padding scheme would pad N bytes of N after the data.
+fn add_padding(data: &mut Vec<u8>) -> Result<(), Ctap2StatusCode> {
+    // The data should be between 1 to MAX_PADDING_LENGTH bytes for the padding scheme to be valid.
+    if data.is_empty() || data.len() > MAX_PADDING_LENGTH as usize {
+        return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    }
+    let pad_length = MAX_PADDING_LENGTH - (data.len() as u8 - 1);
+    data.extend(core::iter::repeat(pad_length).take(pad_length as usize));
+    Ok(())
+}
+
+fn remove_padding(data: &mut Vec<u8>) -> Result<(), Ctap2StatusCode> {
+    if data.len() != MAX_PADDING_LENGTH as usize + 1 {
+        // This is an internal error instead of corrupted credential ID which we should just ignore because
+        // we've already checked that the HMAC matched.
+        return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    }
+    let pad_length = *data.last().unwrap();
+    if pad_length == 0 || pad_length > MAX_PADDING_LENGTH {
+        return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    }
+    data.truncate(data.len() - pad_length as usize);
```

Comment:
Shouldn't we also check that all padding bytes are equal?

Reply:
Done